// es-toolkit/dist/compat/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const initial = require('../_chunk/initial-y0QrPY.js');
const promise_index = require('../_chunk/index-CwRt_M.js');
const function_index = require('../function/index.js');
const math_index = require('../math/index.js');
const randomInt = require('../_chunk/randomInt-CF7bZK.js');
const toMerged = require('../_chunk/toMerged-BGwYW5.js');
const isFunction = require('../_chunk/isFunction-aCEz9d.js');
const isTypedArray$1 = require('../_chunk/isTypedArray-BBEkFl.js');
const string_index = require('../string/index.js');

function chunk(arr, size = 1) {
    size = Math.max(Math.floor(size), 0);
    if (size === 0) {
        return [];
    }
    return initial.chunk(arr, size);
}

function concat(...values) {
    return initial.flatten(values);
}

function difference(arr, ...values) {
    const arr1 = arr;
    const arr2 = initial.flatten(values);
    return initial.difference(arr1, arr2);
}

function fill(array, value, start = 0, end = array.length) {
    start = Math.floor(start);
    end = Math.floor(end);
    if (!start) {
        start = 0;
    }
    if (!end) {
        end = 0;
    }
    return initial.fill(array, value, start, end);
}

const IS_PLAIN = /^\w*$/;
const IS_DEEP = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/;
function isDeepKey(key) {
    switch (typeof key) {
        case 'number':
        case 'symbol': {
            return false;
        }
        case 'string': {
            return !IS_PLAIN.test(key) && IS_DEEP.test(key);
        }
    }
}

function isSymbol(value) {
    return typeof value === 'symbol' || (value != null && value instanceof Symbol);
}

function toKey(value) {
    if (typeof value === 'string' || isSymbol(value)) {
        return value;
    }
    if (Object.is(value?.valueOf(), -0)) {
        return '-0';
    }
    return `${value}`;
}

function toPath(deepKey) {
    const ESCAPE_REGEXP = /\\(\\)?/g;
    const PROPERTY_REGEXP = RegExp('[^.[\\]]+' +
        '|' +
        '\\[(?:' +
        '([^"\'][^[]*)' +
        '|' +
        '(["\'])((?:(?!\\2)[^\\\\]|\\\\.)*?)\\2' +
        ')\\]' +
        '|' +
        '(?=(?:\\.|\\[\\])(?:\\.|\\[\\]|$))', 'g');
    const result = [];
    if (deepKey[0] === '.') {
        result.push('');
    }
    let match;
    let lastIndex = 0;
    while ((match = PROPERTY_REGEXP.exec(deepKey)) !== null) {
        let key = match[0];
        const expr = match[1];
        const quote = match[2];
        const substr = match[3];
        if (quote) {
            key = substr.replace(ESCAPE_REGEXP, '$1');
        }
        else if (expr) {
            key = expr;
        }
        result.push(key);
        if (PROPERTY_REGEXP.lastIndex === lastIndex) {
            PROPERTY_REGEXP.lastIndex++;
        }
        else {
            lastIndex = PROPERTY_REGEXP.lastIndex;
        }
    }
    return result;
}

function get(object, path, defaultValue) {
    let resolvedPath;
    if (Array.isArray(path)) {
        resolvedPath = path;
    }
    else if (typeof path === 'string' && isDeepKey(path) && object?.[path] == null) {
        resolvedPath = toPath(path);
    }
    else {
        resolvedPath = [path];
    }
    if (resolvedPath.length === 0) {
        return defaultValue;
    }
    let current = object;
    let index;
    for (index = 0; index < resolvedPath.length && current != null; index++) {
        const key = toKey(resolvedPath[index]);
        current = current[key];
    }
    if (current === null && index === resolvedPath.length) {
        return current;
    }
    return current ?? defaultValue;
}

function property(path) {
    return function (object) {
        return get(object, path);
    };
}

function isArrayMatch(target, source) {
    if (source.length === 0) {
        return true;
    }
    if (!Array.isArray(target)) {
        return false;
    }
    const countedIndex = new Set();
    for (let i = 0; i < source.length; i++) {
        const sourceItem = source[i];
        const index = target.findIndex((targetItem, index) => {
            return isMatch(targetItem, sourceItem) && !countedIndex.has(index);
        });
        if (index === -1) {
            return false;
        }
        countedIndex.add(index);
    }
    return true;
}

function isMapMatch(target, source) {
    if (source.size === 0) {
        return true;
    }
    if (!(target instanceof Map)) {
        return false;
    }
    for (const [key, value] of source.entries()) {
        if (!isMatch(target.get(key), value)) {
            return false;
        }
    }
    return true;
}

function isSetMatch(target, source) {
    if (source.size === 0) {
        return true;
    }
    if (!(target instanceof Set)) {
        return false;
    }
    return isArrayMatch([...target], [...source]);
}

function isMatch(target, source) {
    if (source === target) {
        return true;
    }
    switch (typeof source) {
        case 'object': {
            if (source == null) {
                return true;
            }
            const keys = Object.keys(source);
            if (target == null) {
                if (keys.length === 0) {
                    return true;
                }
                return false;
            }
            if (Array.isArray(source)) {
                return isArrayMatch(target, source);
            }
            if (source instanceof Map) {
                return isMapMatch(target, source);
            }
            if (source instanceof Set) {
                return isSetMatch(target, source);
            }
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                if (!isTypedArray$1.isPrimitive(target) && !(key in target)) {
                    return false;
                }
                if (source[key] === undefined && target[key] !== undefined) {
                    return false;
                }
                if (!isMatch(target[key], source[key])) {
                    return false;
                }
            }
            return true;
        }
        case 'function': {
            if (Object.keys(source).length > 0) {
                return isMatch(target, { ...source });
            }
            return false;
        }
        default: {
            return !source;
        }
    }
}

function matches(source) {
    source = toMerged.cloneDeep(source);
    return (target) => {
        return isMatch(target, source);
    };
}

function cloneDeep(obj) {
    if (typeof obj !== 'object') {
        return toMerged.cloneDeep(obj);
    }
    switch (Object.prototype.toString.call(obj)) {
        case isFunction.numberTag:
        case isFunction.stringTag:
        case isFunction.booleanTag: {
            const result = new obj.constructor(obj?.valueOf());
            toMerged.copyProperties(result, obj);
            return result;
        }
        case isFunction.argumentsTag: {
            const result = {};
            toMerged.copyProperties(result, obj);
            result.length = obj.length;
            result[Symbol.iterator] = obj[Symbol.iterator];
            return result;
        }
        default: {
            return toMerged.cloneDeep(obj);
        }
    }
}

const IS_UNSIGNED_INTEGER = /^(?:0|[1-9]\d*)$/;
function isIndex(value) {
    switch (typeof value) {
        case 'number': {
            return Number.isInteger(value) && value >= 0 && value < Number.MAX_SAFE_INTEGER;
        }
        case 'symbol': {
            return false;
        }
        case 'string': {
            return IS_UNSIGNED_INTEGER.test(value);
        }
    }
}

function isArguments(value) {
    return value !== null && typeof value === 'object' && isFunction.getTag(value) === '[object Arguments]';
}

function has(object, path) {
    let resolvedPath;
    if (Array.isArray(path)) {
        resolvedPath = path;
    }
    else if (typeof path === 'string' && isDeepKey(path) && object?.[path] == null) {
        resolvedPath = toPath(path);
    }
    else {
        resolvedPath = [path];
    }
    if (resolvedPath.length === 0) {
        return false;
    }
    let current = object;
    for (let i = 0; i < resolvedPath.length; i++) {
        const key = resolvedPath[i];
        if (current == null || !Object.prototype.hasOwnProperty.call(current, key)) {
            const isSparseIndex = (Array.isArray(current) || isArguments(current)) && isIndex(key) && key < current.length;
            if (!isSparseIndex) {
                return false;
            }
        }
        current = current[key];
    }
    return true;
}

function matchesProperty(property, source) {
    property = Array.isArray(property) ? property : toKey(property);
    source = cloneDeep(source);
    return function (target) {
        const result = get(target, property);
        if (result === undefined) {
            return has(target, property);
        }
        if (source === undefined) {
            return result === undefined;
        }
        return isMatch(result, source);
    };
}

function find(source, doesMatch) {
    let values = source;
    if (!Array.isArray(source)) {
        values = Object.values(source);
    }
    switch (typeof doesMatch) {
        case 'function': {
            if (!Array.isArray(source)) {
                const entries = Object.entries(source);
                for (let i = 0; i < entries.length; i++) {
                    const entry = entries[i];
                    const key = entry[0];
                    const value = entry[1];
                    if (doesMatch(value, key, source)) {
                        return value;
                    }
                }
                return undefined;
            }
            return values.find(doesMatch);
        }
        case 'object': {
            if (Array.isArray(doesMatch) && doesMatch.length === 2) {
                const key = doesMatch[0];
                const value = doesMatch[1];
                return values.find(matchesProperty(key, value));
            }
            else {
                return values.find(matches(doesMatch));
            }
        }
        case 'string': {
            return values.find(property(doesMatch));
        }
    }
}

function findIndex(source, doesMatch) {
    switch (typeof doesMatch) {
        case 'function': {
            return source.findIndex(doesMatch);
        }
        case 'object': {
            if (Array.isArray(doesMatch) && doesMatch.length === 2) {
                const key = doesMatch[0];
                const value = doesMatch[1];
                return source.findIndex(matchesProperty(key, value));
            }
            else {
                return source.findIndex(matches(doesMatch));
            }
        }
        case 'string': {
            return source.findIndex(property(doesMatch));
        }
    }
}

function flatten(value, depth = 1) {
    const result = [];
    const flooredDepth = Math.floor(depth);
    if (!Array.isArray(value)) {
        return result;
    }
    const recursive = (arr, currentDepth) => {
        for (const item of arr) {
            if (currentDepth < flooredDepth &&
                (Array.isArray(item) ||
                    Boolean(item?.[Symbol.isConcatSpreadable]) ||
                    (item !== null && typeof item === 'object' && Object.prototype.toString.call(item) === '[object Arguments]'))) {
                if (Array.isArray(item)) {
                    recursive(item, currentDepth + 1);
                }
                else {
                    recursive(Array.from(item), currentDepth + 1);
                }
            }
            else {
                result.push(item);
            }
        }
    };
    recursive(value, 0);
    return result;
}

function flattenDeep(value) {
    return flatten(value, Infinity);
}

function flattenDepth(value, depth = 1) {
    return flatten(value, depth);
}

const regexIsDeepProp = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/;
const regexIsPlainProp = /^\w*$/;
function isKey(value, object) {
    if (Array.isArray(value)) {
        return false;
    }
    if (typeof value === 'number' || typeof value === 'boolean' || value == null || isSymbol(value)) {
        return true;
    }
    return ((typeof value === 'string' && (regexIsPlainProp.test(value) || !regexIsDeepProp.test(value))) ||
        (object != null && Object.hasOwn(object, value)));
}

function getPath(key, object) {
    if (Array.isArray(key)) {
        const path = [];
        for (let i = 0; i < key.length; i++) {
            const k = key[i];
            if (isKey(k, object)) {
                object = object[k];
                path.push(k);
            }
            else {
                const keys = toPath(k);
                for (let i = 0; i < keys.length; i++) {
                    object = object[keys[i]];
                    path.push(keys[i]);
                }
            }
        }
        return path;
    }
    return isKey(key, object) ? key : toPath(key);
}

function orderBy(collection, keys, orders) {
    if (collection == null) {
        return [];
    }
    if (!Array.isArray(keys)) {
        keys = keys == null ? [] : [keys];
    }
    if (!Array.isArray(orders)) {
        orders = orders == null ? [] : [orders];
    }
    const compareValues = (a, b, order) => {
        if (a < b) {
            return order === 'desc' ? 1 : -1;
        }
        if (a > b) {
            return order === 'desc' ? -1 : 1;
        }
        return 0;
    };
    const getValueByPath = (key, obj) => {
        if (Array.isArray(key)) {
            let value = obj;
            for (let i = 0; i < key.length; i++) {
                value = value[key[i]];
            }
            return value;
        }
        return obj[key];
    };
    keys = keys.map(key => getPath(key, collection[0]));
    const shallowCopiedCollection = collection.slice();
    const orderedCollection = shallowCopiedCollection.sort((a, b) => {
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            const valueA = getValueByPath(key, a);
            const valueB = getValueByPath(key, b);
            const order = String(orders[i]);
            const comparedResult = compareValues(valueA, valueB, order);
            if (comparedResult !== 0) {
                return comparedResult;
            }
        }
        return 0;
    });
    return orderedCollection;
}

function size(target) {
    if (isFunction.isNil(target)) {
        return 0;
    }
    if (target instanceof Map || target instanceof Set) {
        return target.size;
    }
    return Object.keys(target).length;
}

function set(obj, path, value) {
    const resolvedPath = Array.isArray(path) ? path : typeof path === 'string' ? toPath(path) : [path];
    let current = obj;
    for (let i = 0; i < resolvedPath.length - 1; i++) {
        const key = resolvedPath[i];
        const nextKey = resolvedPath[i + 1];
        if (current[key] == null) {
            current[key] = isIndex(nextKey) ? [] : {};
        }
        current = current[key];
    }
    const lastKey = resolvedPath[resolvedPath.length - 1];
    current[lastKey] = value;
    return obj;
}

function zipObjectDeep(keys, values) {
    const result = {};
    const zipped = initial.zip(keys, values);
    for (let i = 0; i < zipped.length; i++) {
        const [key, value] = zipped[i];
        if (key != null) {
            set(result, key, value);
        }
    }
    return result;
}

function indexOf(array, searchElement, fromIndex) {
    if (array == null) {
        return -1;
    }
    if (Number.isNaN(searchElement)) {
        fromIndex = fromIndex ?? 0;
        if (fromIndex < 0) {
            fromIndex = Math.max(0, array.length + fromIndex);
        }
        for (let i = fromIndex; i < array.length; i++) {
            if (Number.isNaN(array[i])) {
                return i;
            }
        }
        return -1;
    }
    return array.indexOf(searchElement, fromIndex);
}

function ary(func, n = func.length, guard) {
    if (guard) {
        n = func.length;
    }
    if (Number.isNaN(n) || n < 0) {
        n = 0;
    }
    return function_index.ary(func, n);
}

function bind(func, thisObj, ...partialArgs) {
    const binded = function (...providedArgs) {
        const args = [];
        let startIndex = 0;
        for (let i = 0; i < partialArgs.length; i++) {
            const arg = partialArgs[i];
            if (arg === bind.placeholder) {
                args.push(providedArgs[startIndex++]);
            }
            else {
                args.push(arg);
            }
        }
        for (let i = startIndex; i < providedArgs.length; i++) {
            args.push(providedArgs[i]);
        }
        if (this instanceof binded) {
            return new func(...args);
        }
        return func.apply(thisObj, args);
    };
    return binded;
}
const bindPlaceholder = Symbol('bind.placeholder');
bind.placeholder = bindPlaceholder;

function rest(func, start = func.length - 1) {
    start = Number.parseInt(start, 10);
    if (Number.isNaN(start) || start < 0) {
        start = func.length - 1;
    }
    return function_index.rest(func, start);
}

function identity(x) {
    return x;
}

function mapKeys(object, getNewKey) {
    getNewKey = getNewKey ?? identity;
    switch (typeof getNewKey) {
        case 'string':
        case 'symbol':
        case 'number':
        case 'object': {
            return toMerged.mapKeys(object, property(getNewKey));
        }
        case 'function': {
            return toMerged.mapKeys(object, getNewKey);
        }
    }
}

function mapValues(object, getNewValue) {
    getNewValue = getNewValue ?? identity;
    switch (typeof getNewValue) {
        case 'string':
        case 'symbol':
        case 'number':
        case 'object': {
            return toMerged.mapValues(object, property(getNewValue));
        }
        case 'function': {
            return toMerged.mapValues(object, getNewValue);
        }
    }
}

function isPlainObject(object) {
    if (typeof object !== 'object') {
        return false;
    }
    if (object == null) {
        return false;
    }
    if (Object.getPrototypeOf(object) === null) {
        return true;
    }
    if (Object.prototype.toString.call(object) !== '[object Object]') {
        const tag = object[Symbol.toStringTag];
        if (tag == null) {
            return false;
        }
        const isTagReadonly = !Object.getOwnPropertyDescriptor(object, Symbol.toStringTag)?.writable;
        if (isTagReadonly) {
            return false;
        }
        return object.toString() === `[object ${tag}]`;
    }
    let proto = object;
    while (Object.getPrototypeOf(proto) !== null) {
        proto = Object.getPrototypeOf(proto);
    }
    return Object.getPrototypeOf(object) === proto;
}

function isTypedArray(x) {
    return isTypedArray$1.isTypedArray(x);
}

function mergeWith(object, ...otherArgs) {
    const sources = otherArgs.slice(0, -1);
    const merge = otherArgs[otherArgs.length - 1];
    let result = object;
    for (let i = 0; i < sources.length; i++) {
        const source = sources[i];
        result = mergeWithDeep(object, source, merge, new Map());
    }
    return result;
}
function mergeWithDeep(target, source, merge, stack) {
    if (source == null || typeof source !== 'object') {
        return target;
    }
    if (stack.has(source)) {
        return toMerged.clone(stack.get(source));
    }
    stack.set(source, target);
    if (Array.isArray(source)) {
        source = source.slice();
        for (let i = 0; i < source.length; i++) {
            source[i] = source[i] ?? undefined;
        }
    }
    const sourceKeys = Object.keys(source);
    for (let i = 0; i < sourceKeys.length; i++) {
        const key = sourceKeys[i];
        let sourceValue = source[key];
        let targetValue = target[key];
        if (isArguments(sourceValue)) {
            sourceValue = { ...sourceValue };
        }
        if (isArguments(targetValue)) {
            targetValue = { ...targetValue };
        }
        if (typeof Buffer !== 'undefined' && Buffer.isBuffer(sourceValue)) {
            sourceValue = cloneDeep(sourceValue);
        }
        if (Array.isArray(sourceValue)) {
            targetValue = typeof targetValue === 'object' ? Array.from(targetValue ?? []) : [];
        }
        const merged = merge(targetValue, sourceValue, key, target, source, stack);
        if (merged != null) {
            target[key] = merged;
        }
        else if (Array.isArray(sourceValue)) {
            target[key] = mergeWithDeep(targetValue, sourceValue, merge, stack);
        }
        else if (toMerged.isObjectLike(targetValue) && toMerged.isObjectLike(sourceValue)) {
            target[key] = mergeWithDeep(targetValue, sourceValue, merge, stack);
        }
        else if (targetValue == null && isPlainObject(sourceValue)) {
            target[key] = mergeWithDeep({}, sourceValue, merge, stack);
        }
        else if (targetValue == null && isTypedArray(sourceValue)) {
            target[key] = cloneDeep(sourceValue);
        }
        else if (targetValue === undefined || sourceValue !== undefined) {
            target[key] = sourceValue;
        }
    }
    return target;
}

function merge(object, ...sources) {
    return mergeWith(object, ...sources, function_index.noop);
}

function isArray(value) {
    return Array.isArray(value);
}

function isArrayLike(value) {
    return value != null && typeof value !== 'function' && isFunction.isLength(value.length);
}

function isBoolean(x) {
    if (x === true || x === false) {
        return true;
    }
    if (typeof x === 'object' && x != null && isFunction.getTag(x) === '[object Boolean]') {
        return true;
    }
    return false;
}

function isString(value) {
    if (typeof value === 'string') {
        return true;
    }
    if (typeof value === 'object' && value != null && isFunction.getTag(value) === '[object String]') {
        return true;
    }
    return false;
}

const startsWith = (str, target, position = 0) => {
    return str.startsWith(target, position);
};

const endsWith = (str, target, position = str.length) => {
    return str.endsWith(target, position);
};

function padStart(str, length = 0, chars = ' ') {
    return str.padStart(length, chars);
}

function padEnd(str, length = 0, chars = ' ') {
    return str.padEnd(length, chars);
}

function max(items = []) {
    let maxElement = items[0];
    let max = undefined;
    for (const element of items) {
        if (max == null || element > max) {
            max = element;
            maxElement = element;
        }
    }
    return maxElement;
}

function min(items = []) {
    let minElement = items[0];
    let min = undefined;
    for (const element of items) {
        if (min == null || element < min) {
            min = element;
            minElement = element;
        }
    }
    return minElement;
}

exports.compact = initial.compact;
exports.countBy = initial.countBy;
exports.differenceBy = initial.differenceBy;
exports.differenceWith = initial.differenceWith;
exports.drop = initial.drop;
exports.dropRight = initial.dropRight;
exports.dropRightWhile = initial.dropRightWhile;
exports.dropWhile = initial.dropWhile;
exports.first = initial.head;
exports.flatMap = initial.flatMap;
exports.forEachRight = initial.forEachRight;
exports.groupBy = initial.groupBy;
exports.head = initial.head;
exports.initial = initial.initial;
exports.intersection = initial.intersection;
exports.intersectionBy = initial.intersectionBy;
exports.intersectionWith = initial.intersectionWith;
exports.isSubset = initial.isSubset;
exports.keyBy = initial.keyBy;
exports.last = initial.last;
exports.maxBy = initial.maxBy;
exports.minBy = initial.minBy;
exports.partition = initial.partition;
exports.sample = initial.sample;
exports.sampleSize = initial.sampleSize;
exports.shuffle = initial.shuffle;
exports.sortBy = initial.sortBy;
exports.tail = initial.tail;
exports.take = initial.take;
exports.takeRight = initial.takeRight;
exports.takeRightWhile = initial.takeRightWhile;
exports.takeWhile = initial.takeWhile;
exports.toFilled = initial.toFilled;
exports.union = initial.union;
exports.unionBy = initial.unionBy;
exports.unionWith = initial.unionWith;
exports.uniq = initial.uniq;
exports.uniqBy = initial.uniqBy;
exports.uniqWith = initial.uniqWith;
exports.unzip = initial.unzip;
exports.unzipWith = initial.unzipWith;
exports.without = initial.without;
exports.xor = initial.xor;
exports.xorBy = initial.xorBy;
exports.xorWith = initial.xorWith;
exports.zip = initial.zip;
exports.zipObject = initial.zipObject;
exports.zipWith = initial.zipWith;
exports.AbortError = promise_index.AbortError;
exports.TimeoutError = promise_index.TimeoutError;
exports.delay = promise_index.delay;
exports.withTimeout = promise_index.withTimeout;
exports.after = function_index.after;
exports.before = function_index.before;
exports.debounce = function_index.debounce;
exports.memoize = function_index.memoize;
exports.negate = function_index.negate;
exports.noop = function_index.noop;
exports.once = function_index.once;
exports.partial = function_index.partial;
exports.partialRight = function_index.partialRight;
exports.throttle = function_index.throttle;
exports.unary = function_index.unary;
exports.clamp = math_index.clamp;
exports.inRange = math_index.inRange;
exports.mean = math_index.mean;
exports.meanBy = math_index.meanBy;
exports.range = math_index.range;
exports.round = math_index.round;
exports.sum = math_index.sum;
exports.sumBy = math_index.sumBy;
exports.random = randomInt.random;
exports.randomInt = randomInt.randomInt;
exports.clone = toMerged.clone;
exports.cloneDeep = toMerged.cloneDeep;
exports.flattenObject = toMerged.flattenObject;
exports.invert = toMerged.invert;
exports.isObjectLike = toMerged.isObjectLike;
exports.omit = toMerged.omit;
exports.omitBy = toMerged.omitBy;
exports.pick = toMerged.pick;
exports.pickBy = toMerged.pickBy;
exports.toMerged = toMerged.toMerged;
exports.isEqual = isFunction.isEqual;
exports.isFunction = isFunction.isFunction;
exports.isLength = isFunction.isLength;
exports.isNil = isFunction.isNil;
exports.isNotNil = isFunction.isNotNil;
exports.isNull = isFunction.isNull;
exports.isUndefined = isFunction.isUndefined;
exports.isPrimitive = isTypedArray$1.isPrimitive;
exports.camelCase = string_index.camelCase;
exports.capitalize = string_index.capitalize;
exports.kebabCase = string_index.kebabCase;
exports.lowerCase = string_index.lowerCase;
exports.pascalCase = string_index.pascalCase;
exports.snakeCase = string_index.snakeCase;
exports.startCase = string_index.startCase;
exports.ary = ary;
exports.bind = bind;
exports.chunk = chunk;
exports.concat = concat;
exports.difference = difference;
exports.endsWith = endsWith;
exports.fill = fill;
exports.find = find;
exports.findIndex = findIndex;
exports.flatten = flatten;
exports.flattenDeep = flattenDeep;
exports.flattenDepth = flattenDepth;
exports.get = get;
exports.has = has;
exports.indexOf = indexOf;
exports.isArguments = isArguments;
exports.isArray = isArray;
exports.isArrayLike = isArrayLike;
exports.isBoolean = isBoolean;
exports.isMatch = isMatch;
exports.isPlainObject = isPlainObject;
exports.isString = isString;
exports.isSymbol = isSymbol;
exports.isTypedArray = isTypedArray;
exports.mapKeys = mapKeys;
exports.mapValues = mapValues;
exports.matches = matches;
exports.matchesProperty = matchesProperty;
exports.max = max;
exports.merge = merge;
exports.mergeWith = mergeWith;
exports.min = min;
exports.orderBy = orderBy;
exports.padEnd = padEnd;
exports.padStart = padStart;
exports.property = property;
exports.rest = rest;
exports.set = set;
exports.size = size;
exports.startsWith = startsWith;
exports.zipObjectDeep = zipObjectDeep;
