// es-toolkit/dist/object/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const toMerged = require('../_chunk/toMerged-BGwYW5.js');

function mergeWith(target, source, merge) {
    const sourceKeys = Object.keys(source);
    for (let i = 0; i < sourceKeys.length; i++) {
        const key = sourceKeys[i];
        const sourceValue = source[key];
        const targetValue = target[key];
        const merged = merge(targetValue, sourceValue, key, target, source);
        if (merged != null) {
            target[key] = merged;
        }
        else if (Array.isArray(sourceValue)) {
            target[key] = mergeWith(targetValue ?? [], sourceValue, merge);
        }
        else if (toMerged.isObjectLike(targetValue) && toMerged.isObjectLike(sourceValue)) {
            target[key] = mergeWith(targetValue ?? {}, sourceValue, merge);
        }
        else if (targetValue === undefined || sourceValue !== undefined) {
            target[key] = sourceValue;
        }
    }
    return target;
}

exports.clone = toMerged.clone;
exports.cloneDeep = toMerged.cloneDeep;
exports.flattenObject = toMerged.flattenObject;
exports.invert = toMerged.invert;
exports.mapKeys = toMerged.mapKeys;
exports.mapValues = toMerged.mapValues;
exports.merge = toMerged.merge;
exports.omit = toMerged.omit;
exports.omitBy = toMerged.omitBy;
exports.pick = toMerged.pick;
exports.pickBy = toMerged.pickBy;
exports.toMerged = toMerged.toMerged;
exports.mergeWith = mergeWith;
